diff --git a/torch_scatter/scatter.py b/torch_scatter/scatter.py
index d4d8050..b8db296 100644
--- a/torch_scatter/scatter.py
+++ b/torch_scatter/scatter.py
@@ -1,7 +1,7 @@
 from typing import Optional, Tuple
 
 import torch
-
+import mx_driving
 from .utils import broadcast
 
 
@@ -172,6 +172,7 @@ def scatter(src: torch.Tensor,
     elif reduce == 'min':
         return scatter_min(src, index, dim, out, dim_size)[0]
     elif reduce == 'max':
-        return scatter_max(src, index, dim, out, dim_size)[0]
+        return mx_driving.scatter_max(src.float(),index.to(torch.int32),out)[0]
     else:
         raise ValueError
